pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/IngestionDelayTracker.java

```diff
@@ -83,22 +82,35 @@
  *
  * TODO: handle bug situations like the one where a partition is not allocated to a given server due to a bug.
  */
-
 public class IngestionDelayTracker {
 
   private static class IngestionInfo {
-    final long _ingestionTimeMs;
-    final long _firstStreamIngestionTimeMs;
-    final StreamPartitionMsgOffset _currentOffset;
-    final StreamPartitionMsgOffset _latestOffset;
+    volatile long _ingestionTimeMs;
+    volatile long _firstStreamIngestionTimeMs;
+    volatile StreamPartitionMsgOffset _currentOffset;
+    volatile StreamPartitionMsgOffset _latestOffset;
```

Comment:
Why do we need to make this class mutable instead of allocate new instances?

Reply:
Yeah, my thought process was to not allocate too many small lives objects since this update metric method is called too frequently. 

I can change it to create new instances as well.

Reply:
It is fine to be volatile, but please mark all these attributes as nullable. I know they were not marked as such before, but they should have been marked. BTW, `_latestOffsetFetcher` should also be marked as nullable

Reply:
`volatile` is a bit of Java that I don't yet fully understand: would someone explain what making these 4 fields `volatile` will do within this class?

Reply:
Basically instead of creating new objects, we simply keep on updating the value of current offset and latest offset in this object

We the prometheus actually calls from the metric, we simply pull the last updated values from this object and return it to prom

tbh, this logic has not been changed, it's just that I have marked them as volatile so that the updates to these variables are reflected correctly. These fields should always have been marked as volatile but we don't do that in a lot in our code :P

Reply:
Gotcha, thanks for the explanation. From what I read yesterday, it seems like volatile is to tell the compiler to constrain how it optimizes code that uses these variables so that concurrent execution of this code behaves correctly.